### ISSUE TITLE
AsciiMath input

### ISSUE THREAD

**Issue by smilank3** [KaTeX/KaTeX#1472]
is it possible to support AsciiMath input.

**Comment by edemaine** [comment]
No, KaTeX does not support AsciiMath input.  I think the goal/focus is on just supporting LaTeX input, but others can feel free to chime in.  If there are AsciiMath -> LaTeX converters out there, you could certainly chain them together.  Otherwise, you might want to stick to MathJax...

I'm curious: in what context do you use AsciiMath?  Do you have a lot of existing AsciiMath content, or do you prefer writing in AsciiMath, or something else?

**Comment by smilank3** [comment]
AsciiMath looks great and i think it is easy (to learn and write) compare to latex. 
any suggestion for AsciiMath -> LaTeX  library. Thank you.

**Comment by pkra** [comment]
> any suggestion for AsciiMath -> LaTeX library. Thank you.

This was written with KaTeX in mind: https://github.com/christianp/asciimath2tex

**Comment by smilank3** [comment]
@pkra Thanks it works. 👍

**Comment by edemaine** [comment]
Glad to hear it! Thanks @pkra. I've created #1474 to add a link to this project.

**Comment by ecstrema** [comment]
I'm interested in adding support for ascii math without first transpiling to latex. For that I'd like some clarification on the structure of katex. Here's how it works to my understanding, please correct me if I'm wrong:

```mermaid
graph LR;
    Latex_Input-->AST-->HTML;
```

I'd like to create a similar AST but from ascii math. So it would then look like that:


```mermaid
graph LR;
	AsciiMath_Input-->Own_AST-->transformer-->AST;
    Latex_Input-->AST-->HTML;
```

This means I'd like to know the format of the AST that is sent to the final HTML renderer, in order to produce a similar AST.

As to why:

> I'm curious: in what context do you use AsciiMath? Do you have a lot of existing AsciiMath content, or do you prefer writing in AsciiMath, or something else?

I've been writing latex to this point, and although I got quite fluent with it, I never managed to like it. There are gotchas everywhere, too many things you need to learn by heart, and it stays ugly and unreadable. Modifying long equations is a nightmare and it's overall too verbose. 

When I discovered Ascii math 6 months ago, I realized there was absolutely no point in having such a complicated syntax. Ascii math provides everything in a much simpler, clean and efficient way.

And yes, there are shortcomings, like no proper support for multiline equations, that should be taken care of, but the core things are there and I now really have a hard time coming back to plain ol' latex.

The few hours I could spend on this will probably save myself wayyy more time in the future, not speaking about all the others who would use it, particularly if it's integrated to Katex.

**Comment by ecstrema** [comment]
But MathJax is sloooooow

**Comment by edemaine** [comment]
KaTeX aims to provide LaTeX compatibility, which puts AsciiMath out of scope. Is <https://github.com/christianp/asciimath2tex> + KaTeX insufficient for your purposes?

**Comment by edemaine** [comment]
I think what would make sense would be a `contrib` library that supports AsciiMath via asciimath2tex + KaTeX, via `` \`...\` ``. This would put it at the same level as `contrib/auto-render` which is what processes `$` etc. expressions.

@ecstrema's approach of going directly to our AST is a possibility, and would safe a parsing step. But the KaTeX AST is closely tied to LaTeX, so I'm not sure how ideal it is for this purpose. Then again, seeing how single [asciimath2tex](https://github.com/christianp/asciimath2tex/blob/master/asciimath2tex.js) is, maybe it's fine. The AST is "documented" by each function (see [`src/functions`](https://github.com/KaTeX/KaTeX/tree/main/src/functions)); in each case, the `handler` method produces an AST node (JS object), and `htmlBuilder` and `mathmlBuilder` turn this AST node into the relevant HTML or MathML.

We'd happily take a PR in either direction. I'm going to reopen this issue for it to get more attention.
